doc/kernel/usermode/syscalls.rst

```diff
@@ -0,0 +1,259 @@
+system call and jump into it.
+
+Handler functions only run when system call APIs are invoked from user mode.
+If an API is invoked from supervisor mode, the implementation is simply called.
+
+The purpose of the handler function is to validate all the arguments passed in.
+This includes:
+
+* Any kernel object pointers provided. For example, the semaphore APIs must
+  ensure that the semaphore object passed in is a valid semaphore and that
+  the calling thread has permission on it.
+
+* Any memory buffers passed in from user mode. Checks must be made that the
+  calling thread has read or write permissions on the provided buffer.
+
+* Any other arguments that have a limited range of valid values.
+
+Handler functions involve a great deal of boilerplate code which has been
+made simpler by some macros in ``kernel/include/syscall_handlers.h``. It is
+strongly recommended that any handler function be declared using these macros.
+
+Argument Validation
+-------------------
+
+Several macros exist which are used to validate arguments:
+
+* :c:macro:`_SYSCALL_OBJ()` Checks a memory address to assert that it is
+  a valid kernel object of the expected type, that the calling thread
+  has permissions on it, and that the object is initialized.
+
+* :c:macro:`_SYSCALL_OBJ_INIT()` is the same as
+  :c:macro:`_SYSCALL_OBJ()`, except that the provided object may be
+  uninitialized. This is useful for handlers of object init functions.
+
+* :c:macro:`_SYSCALL_OBJ_NEVER_INIT()` is the same as
+  :c:macro:`_SYSCALL_OBJ()`, except that the provided object must be
+  uninitialized. This is not used very often, currently only for
+  :c:func:`k_thread_create()`.
+
+* :c:macro:`_SYSCALL_MEMORY_READ()` validates a memory buffer of a particular
+  size. The calling thread must have read permissions on the entire buffer.
+
+* :c:macro:`_SYSCALL_MEMORY_WRITE()` is the same as
+  :c:macro:`_SYSCALL_MEMORY_READ()` but the calling thread must additionally
+  have write permissions.
+
+* :c:macro:`_SYSCALL_MEMORY_ARRAY_READ()` validates an array whose total size
+  is expressed as separate arguments for the number of elements and the
+  element size. This macro correctly accounts for multiplication overflow
+  when computing the total size. The calling thread must have read permissions
+  on the total size.
+
+* :c:macro:`_SYSCALL_MEMORY_ARRAY_WRITE()` is the same as
+  :c:macro:`_SYSCALL_MEMORY_ARRAY_READ()` but the calling thread must
+  additionally have write permissions.
+
+* :c:macro:`_SYSCALL_VERIFY_MSG()` does a runtime check of some boolean
+  expression which must evaluate to true otherwise the check will fail.
+  A variant :c:macro:`_SYSCALL_VERIFY` exists which does not take
+  a message parameter, instead printing the expression tested if it
+  fails. The latter should only be used for the most obvious of tests.
+
+If any checks fail, a kernel oops will be triggered which will kill the
```

Comment:
change to "If any check fails, a kernel oops will be triggered that will kill the"